filters/shedder/admission_test.go

```diff
@@ -0,0 +1,463 @@
+		return routes[i].Id < routes[j].Id
+	})
+	for _, r := range routes {
+		for _, f := range r.Filters {
+			if ac, ok := f.Filter.(*admissionControl); ok {
+				found = true
+
+				closed := false
+				if vpp.oldFilter != nil {
+					closed = vpp.oldFilter.closed
+				}
+				go func(id string, b bool) {
+					vpp.ch <- tuple{id: id, closed: closed}
+				}(r.Id, closed)
+
+				vpp.oldFilter = ac
+			}
+		}
+	}
+
+	if !found {
+		go func() { vpp.ch <- tuple{id: "", closed: vpp.oldFilter.closed} }()
+	}
+	return routes
+}
+
+func TestAdmissionControlCleanupMultipleFilters(t *testing.T) {
+	for _, ti := range []struct {
+		msg string
+		doc string
+	}{{
+		msg: "no filter",
+		doc: `* -> "%s"`,
+	}, {
+		msg: "one not matching filter",
+		doc: `* -> setRequestHeader("Foo", "bar") -> "%s"`,
+	}, {
+		msg: "one matching filter",
+		doc: `* -> admissionControl("app", "active", "1s", 5, 10, 0.1, 0.95, 0.5) -> "%s"`,
+	}, {
+		msg: "two matching filters",
+		doc: `* -> admissionControl("app", "active", "1s", 5, 10, 0.1, 0.95, 0.5) -> admissionControl("app2", "active", "1s", 5, 10, 0.1, 0.95, 0.5) -> "%s"`,
+	}, {
+		msg: "multi filter with multiple matching filters",
+		doc: `* -> setRequestHeader("Foo", "bar") -> admissionControl("app", "active", "1s", 5, 10, 0.1, 0.95, 0.5) -> status(200) -> admissionControl("app2", "active", "1s", 5, 10, 0.1, 0.95, 0.5) -> setRequestHeader("Foo2", "bar2") -> admissionControl("app3", "active", "1s", 5, 10, 0.1, 0.95, 0.5) -> setRequestHeader("Foo3", "bar3") -> "%s"`,
+	}} {
+		t.Run(ti.msg, func(t *testing.T) {
+			ch := make(chan []*routing.Route)
+			validationProcessor := &validationPostProcessorNumberOfFilters{
+				ch: ch,
+			}
+
+			backend1 := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+				w.WriteHeader(http.StatusOK)
+			}))
+
+			fspec := NewAdmissionControl(Options{})
+			spec, ok := fspec.(*AdmissionControlSpec)
+			if !ok {
+				t.Fatal("FilterSpec is not a AdmissionControlSpec")
+			}
+			preProcessor := spec.PreProcessor()
+			postProcessor := spec.PostProcessor()
+
+			r, err := eskip.Parse(fmt.Sprintf(ti.doc, backend1.URL))
+			if err != nil {
+				t.Fatalf("Failed to parse doc: %v", err)
+			}
+
+			fr := make(filters.Registry)
+			fr.Register(fspec)
+			fr.Register(builtin.NewSetRequestHeader())
+			fr.Register(builtin.NewStatus())
+
+			dc := testdataclient.New(r)
+			proxy := proxytest.WithRoutingOptions(fr, routing.Options{
+				DataClients:    []routing.DataClient{dc},
+				PostProcessors: []routing.PostProcessor{postProcessor, validationProcessor},
+				PreProcessors:  []routing.PreProcessor{preProcessor},
+			}, r...)
+			defer proxy.Close()
+
+			dc.Update(r, nil)
+
+			result := <-ch
+			if result == nil {
+				t.Error("Failed to cleanup filters correctly, found more than one admissionControl filter in one route")
+			}
+		})
+	}
+}
+
+type validationPostProcessorNumberOfFilters struct {
+	ch chan []*routing.Route
```

Comment:
minor: Maybe it could receive `*testing.T` and call `t.Error` if admission filter counter is more than one

Reply:
The problem is that t is not Thread safe. At least in my tests I got warnings or a panic, don't remember anymore.